Source/utils/CallFunctionOverLevels.hpp

```diff
@@ -0,0 +1,92 @@
+/* GRChombo
+ * Copyright 2012 The GRChombo collaboration.
+ * Please refer to LICENSE in GRChombo's root directory.
+ */
+
+#ifndef CALL_FUCNTION_OVER_LEVELS
+#define CALL_FUCNTION_OVER_LEVELS
+
+#include "AMR.H"
+#include "GRAMRLevel.hpp"
+#include "Scheduler.H"
+
+#include <limits> // std::numeric_limits
+
+//! This is just an interface for the AMR scheduler to call some GRAMRLevel (or
+//! any other Example specific level) function on every AMRLevel
+//! Satisfies syntax of Chmbo's Scheduler such that it can be passed to GRAMR
+//! and be scheduled
+template <class Level = GRAMRLevel>
+class CallFunctionOverLevels : public Scheduler::PeriodicFunction
+{
+    bool m_reverse_levels;
+    std::function<void(Level *)> m_func;
+
+    // Use default condstructor of PeriodicFunction
+    using PeriodicFunction::PeriodicFunction;
+
+    AMR *m_amr_ptr; //! pointer to AMR object
+
+  public:
+    CallFunctionOverLevels(std::function<void(Level *)> a_func,
+                           bool a_reverse_levels = true)
+        : m_func(a_func), m_reverse_levels(a_reverse_levels)
+    {
+    }
+
+    // required from Scheduler::PeriodicFunction
+    virtual void setUp(AMR &a_AMR, int a_interval = -1) override
+    {
+        m_amr_ptr = &a_AMR;
+    }
+
+    // required from Scheduler::PeriodicFunction
+    virtual void operator()(int a_step = 0, Real a_time = 0.) override
+    {
+        auto amr_level_ptrs = m_amr_ptr->getAMRLevels().stdVector();
+
+        // need to reverse this vector so that m_func is called in order of
+        // finest level to coarsest. This is important for example for
+        // 'specificPostTimeStep', which is always run in reverse order of
+        // levels
+        if (m_reverse_levels)
+            std::reverse(std::begin(amr_level_ptrs), std::end(amr_level_ptrs));
+
+        for (AMRLevel *amr_level_ptr : amr_level_ptrs)
+            m_func(Level::gr_cast(amr_level_ptr));
+    }
+};
+
+//! This is just an interface for the AMR scheduler to call some Level
+//! function on every AMRLevel
+//! This can either be called directly by calling execute, or passed to an AMR
+//! (as GRAMR) by doing gr_amr.schedule(me) (this version will make it be called
+//! only after plot files are written, if that is ever an interest)
+template <class Level = GRAMRLevel>
+class CallFunctionOverLevelsPtr : public RefCountedPtr<Scheduler>
+{
+    RefCountedPtr<CallFunctionOverLevels<Level>> m_ptr_to_func;
+
+  public:
+    //! interval defines the frequency with which the scheduler will be called
+    //! if added to an AMR
+    CallFunctionOverLevelsPtr(std::function<void(Level *)> a_func,
+                              bool a_reverse_levels = true,
+                              int a_interval = std::numeric_limits<int>::max())
```

Comment:
This is fine, but does setting it to zero not turn it off? Usually in Chombo this works.

Reply:
Other intervals have conditions as `m_plot_interval > 0`, but `Scheduler::execute` just does a `(a_step % interval) == 0`, so if interval is 0 it's probably problematic.